File: dashboard-shiny.R
library(shiny)
library(bslib)
library(bsicons)
library(ggplot2)
library(plotly)
library(dplyr)

# Cargar datos
# Asumiendo que tips.csv está en la misma carpeta
tips <- read.csv("tips.csv")

# Pre-cálculo para el slider
bill_rng <- list(
  min = min(tips$total_bill, na.rm = TRUE),
  max = max(tips$total_bill, na.rm = TRUE)
)

# UI (Interfaz de usuario)
ui <- page_sidebar(
  title = "Restaurant tipping",
  fillable = TRUE,
  
  sidebar = sidebar(
    sliderInput(
      "total_bill",
      "Bill amount",
      min = bill_rng$min,
      max = bill_rng$max,
      value = c(bill_rng$min, bill_rng$max),
      pre = "$"
    ),
    checkboxGroupInput(
      "time",
      "Food service",
      choices = c("Lunch", "Dinner"),
      selected = c("Lunch", "Dinner"),
      inline = TRUE
    ),
    actionButton("reset", "Reset filter", class = "btn-outline-secondary")
  ),
  
  # Fila de Value Boxes
  layout_columns(
    value_box(
      title = "Total tippers",
      value = textOutput("total_tippers"),
      showcase = bs_icon("person")
    ),
    value_box(
      title = "Average bill",
      value = textOutput("average_bill"),
      showcase = bs_icon("currency-dollar")
    )
  ),
  
  # Fila de Tabla y Gráfico
  layout_columns(
    card(
      card_header("Tips data"),
      tableOutput("table"),
      full_screen = TRUE
    ),
    card(
      card_header(
        "Total bill vs tip",
        popover(
          bs_icon("three-dots"),
          radioButtons(
            "scatter_color",
            NULL,
            choices = c("none", "sex", "smoker", "day", "time"),
            inline = TRUE
          ),
          title = "Add a color variable",
          placement = "top"
        ),
      ),
      plotlyOutput("scatterplot")
    )
  )
)

# Server (Lógica)
server <- function(input, output, session) {
  
  # Reactivo para filtrar datos (el equivalente a tips_data())
  tips_filtered <- reactive({
    tips %>%
      filter(
        total_bill >= input$total_bill[1],
        total_bill <= input$total_bill[2],
        time %in% input$time
      )
  })
  
  output$total_tippers <- renderText({
    nrow(tips_filtered())
  })
  
  output$average_bill <- renderText({
    d <- mean(tips_filtered()$total_bill, na.rm = TRUE)
    if (!is.nan(d)) {
      paste0("$", format(round(d, 2)))
    } else {
      "N/A"
    }
  })
  
  output$table <- renderTable({
    tips_filtered()
  })
  
  output$scatterplot <- renderPlotly({
    color_var <- if (input$scatter_color == "none") NULL else input$scatter_color
    
    p <- ggplot(tips_filtered(), aes(x = total_bill, y = tip, color = .data[[color_var]])) +
      geom_point() +
      geom_smooth(method = "loess", formula = y ~ x) +
      theme_minimal()
    
    # Si no hay color, quitamos la leyenda de 'NULL'
    if (input$scatter_color == "none") {
      p <- ggplot(tips_filtered(), aes(x = total_bill, y = tip)) +
        geom_point() +
        geom_smooth(method = "loess", formula = y ~ x) +
        theme_minimal()
    }
    
    ggplotly(p)
  })
  
  # Evento para resetear filtros
  observeEvent(input$reset, {
    updateSliderInput(session, "total_bill", value = c(bill_rng$min, bill_rng$max))
    updateCheckboxGroupInput(session, "time", selected = c("Lunch", "Dinner"))
  })
}

shinyApp(ui, server)

File: dashboard-shiny.py
# Librerias
from pathlib import Path
import faicons as fa
import plotly.express as px
import pandas as pd

from shiny import App, reactive, render, ui
from shinywidgets import output_widget, render_plotly

# directorio de la app
app_dir = Path(__file__).parent

# cargar los datos del tablero
tips = pd.read_csv(app_dir / "tips.csv")

bill_rng = (
    tips["total_bill"].min(),
    tips["total_bill"].max()
)

# UI (Interfaz de usuario)
app_ui = ui.page_sidebar(
    ui.sidebar(
        ui.input_slider(
            "total_bill",
            "Bill amount",
            min=bill_rng[0],
            max=bill_rng[1],
            value=bill_rng,
            pre="$",
        ),
        ui.input_checkbox_group(
            "time",
            "Food service",
            ["Lunch", "Dinner"],
            selected=["Lunch", "Dinner"],
            inline=True,
        ),
        ui.input_action_button("reset", "Reset filter", _class = "btn-outline-secondary")
    ),
    # Fila de Value Boxes
    ui.layout_columns(
        ui.value_box(
            "Total tippers", ui.output_ui("total_tippers"), showcase=fa.icon_svg("user", "regular")
        ),
        ui.value_box(
            "Average bill", ui.output_ui("average_bill"),  showcase=fa.icon_svg("dollar-sign")
        )
    ),
    # Fila de Tabla y Gráfico
    ui.layout_columns(
        ui.card(
            ui.card_header("Tips data"), ui.output_data_frame("table"), full_screen=True
        ),
        ui.card(
            ui.card_header(
                "Total bill vs tip",
                ui.popover(
                    fa.icon_svg("ellipsis"),
                    ui.input_radio_buttons(
                        "scatter_color",
                        None,
                        ["none", "sex", "smoker", "day", "time"],
                        inline=True,
                    ),
                    title="Add a color variable",
                    placement="top",
                ),
            ),
            output_widget("scatterplot")
        )
    ),
    title="Restaurant tipping",
    fillable=True
)

# Server (Lógica)
def server(input, output, session):
    @reactive.calc
    def tips_data():
        bill = input.total_bill()
        return tips[
                    (tips["total_bill"].between(bill[0], bill[1])) & 
                    (tips["time"].isin(input.time()))
                ]

    @render.ui
    def total_tippers():
        return tips_data().shape[0]

    @render.ui
    def average_bill():
        d = tips_data()["total_bill"].mean()
        if d:
            return f"${d:.2f}"
        else:
            return "N/A"

    @render.data_frame
    def table():
        return render.DataGrid(tips_data())

    @render_plotly
    def scatterplot():
        color = input.scatter_color()
        return px.scatter(
            tips_data(),
            x="total_bill",
            y="tip",
            color=None if color == "none" else color,
            trendline="lowess",
        )

    @reactive.effect
    @reactive.event(input.reset)
    def _():
        ui.update_slider("total_bill", value=bill_rng)
        ui.update_checkbox_group("time", selected=["Lunch", "Dinner"])

app = App(app_ui, server)
